File: IAIso-v5.0/core/iaiso-csharp/src/Iaiso.Conformance/ConsentRunner.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text;
using System.Text.Json.Nodes;
using Iaiso.Consent;

namespace Iaiso.Conformance;

internal static class ConsentRunner
{
    public static List<VectorResult> Run(string specRoot)
    {
        var path = Path.Combine(specRoot, "consent", "vectors.json");
        var doc = JsonNode.Parse(File.ReadAllBytes(path))!.AsObject();
        string sharedKey = doc["hs256_key_shared"]!.GetValue<string>();

        var results = new List<VectorResult>();
        if (doc["scope_match"] is JsonArray sm)
            foreach (var v in sm) results.Add(RunScopeMatch(v!.AsObject()));
        if (doc["scope_match_errors"] is JsonArray sme)
            foreach (var v in sme) results.Add(RunScopeMatchError(v!.AsObject()));
        if (doc["valid_tokens"] is JsonArray vt)
            foreach (var v in vt) results.Add(RunValidToken(sharedKey, v!.AsObject()));
        if (doc["invalid_tokens"] is JsonArray it)
            foreach (var v in it) results.Add(RunInvalidToken(sharedKey, v!.AsObject()));
        if (doc["issue_and_verify_roundtrip"] is JsonArray rt)
            foreach (var v in rt) results.Add(RunRoundtrip(sharedKey, v!.AsObject()));
        return results;
    }

    private static VectorResult RunScopeMatch(JsonObject v)
    {
        string name = "scope_match/" + v["name"]!.GetValue<string>();
        var granted = new List<string>();
        foreach (var g in v["granted"]!.AsArray()) granted.Add(g!.GetValue<string>());
        string requested = v["requested"]!.GetValue<string>();
        bool expected = v["expected"]!.GetValue<bool>();
        try
        {
            bool got = Scopes.Granted(granted, requested);
            return got == expected
                ? VectorResult.Pass("consent", name)
                : VectorResult.Fail("consent", name, $"got {got}, want {expected}");
        }
        catch (Exception e)
        {
            return VectorResult.Fail("consent", name, "unexpected exception: " + e.Message);
        }
    }

    private static VectorResult RunScopeMatchError(JsonObject v)
    {
        string name = "scope_match_errors/" + v["name"]!.GetValue<string>();
        var granted = new List<string>();
        foreach (var g in v["granted"]!.AsArray()) granted.Add(g!.GetValue<string>());
        string requested = v["requested"]!.GetValue<string>();
        string expectErr = v["expect_error"]!.GetValue<string>();
        try
        {
            Scopes.Granted(granted, requested);
            return VectorResult.Fail("consent", name,
                $"expected error containing '{expectErr}', got Ok");
        }
        catch (ArgumentException e)
        {
            string msg = e.Message.ToLowerInvariant();
            return msg.Contains(expectErr.ToLowerInvariant())
                ? VectorResult.Pass("consent", name)
                : VectorResult.Fail("consent", name,
                    $"expected error containing '{expectErr}', got: {e.Message}");
        }
    }

    private static Algorithm ParseAlg(JsonObject v)
    {
        if (v["algorithm"] is JsonValue av && av.TryGetValue<string>(out var s))
        {
            try { return AlgorithmExtensions.ParseAlgorithm(s); } catch { }
        }
        return Algorithm.HS256;
    }

    private static string? StringOrDefault(JsonObject v, string key, string? dflt)
    {
        if (v[key] is JsonValue jv && jv.TryGetValue<string>(out var s)) return s;
        return dflt;
    }

    private static VectorResult RunValidToken(string sharedKey, JsonObject v)
    {
        string name = "valid_tokens/" + v["name"]!.GetValue<string>();
        long now = v["now"]!.GetValue<long>();
        string issuer = StringOrDefault(v, "issuer", "iaiso") ?? "iaiso";
        var alg = ParseAlg(v);
        string token = v["token"]!.GetValue<string>();
        var expected = v["expected"]!.AsObject();

        var verifier = Verifier.CreateBuilder()
            .WithHsKey(Encoding.UTF8.GetBytes(sharedKey))
            .WithAlgorithm(alg)
            .WithIssuer(issuer)
            .WithClock(() => now)
            .Build();
        try
        {
            var s = verifier.Verify(token, null);
            string wantSub = expected["sub"]!.GetValue<string>();
            if (s.Subject != wantSub)
                return VectorResult.Fail("consent", name, $"sub: got {s.Subject}, want {wantSub}");
            string wantJti = expected["jti"]!.GetValue<string>();
            if (s.Jti != wantJti)
                return VectorResult.Fail("consent", name, $"jti: got {s.Jti}, want {wantJti}");
            var wantScopes = new List<string>();
            foreach (var n in expected["scopes"]!.AsArray()) wantScopes.Add(n!.GetValue<string>());
            if (!ListEquals(s.Scopes, wantScopes))
                return VectorResult.Fail("consent", name, "scopes mismatch");
            string? wantExec = expected["execution_id"] is JsonValue ev
                && ev.TryGetValue<string>(out var es) ? es : null;
            if (wantExec is null ? s.ExecutionId is not null : wantExec != s.ExecutionId)
                return VectorResult.Fail("consent", name,
                    $"execution_id: got {s.ExecutionId}, want {wantExec}");
            return VectorResult.Pass("consent", name);
        }
        catch (Exception e)
        {
            return VectorResult.Fail("consent", name, "verify failed: " + e.Message);
        }
    }

    private static VectorResult RunInvalidToken(string sharedKey, JsonObject v)
    {
        string name = "invalid_tokens/" + v["name"]!.GetValue<string>();
        long now = v["now"]!.GetValue<long>();
        string issuer = StringOrDefault(v, "issuer", "iaiso") ?? "iaiso";
        var alg = ParseAlg(v);
        string token = v["token"]!.GetValue<string>();
        string? execId = StringOrDefault(v, "execution_id", null);
        string expectErr = v["expect_error"]!.GetValue<string>();

        var verifier = Verifier.CreateBuilder()
            .WithHsKey(Encoding.UTF8.GetBytes(sharedKey))
            .WithAlgorithm(alg)
            .WithIssuer(issuer)
            .WithClock(() => now)
            .Build();
        try
        {
            verifier.Verify(token, execId);
            return VectorResult.Fail("consent", name,
                $"expected error '{expectErr}', got Ok");
        }
        catch (ExpiredTokenException)
        {
            return expectErr == "expired"
                ? VectorResult.Pass("consent", name)
                : VectorResult.Fail("consent", name, $"expected '{expectErr}', got expired");
        }
        catch (RevokedTokenException)
        {
            return expectErr == "revoked"
                ? VectorResult.Pass("consent", name)
                : VectorResult.Fail("consent", name, $"expected '{expectErr}', got revoked");
        }
        catch (InvalidTokenException e)
        {
            return expectErr == "invalid"
                ? VectorResult.Pass("consent", name)
                : VectorResult.Fail("consent", name,
                    $"expected '{expectErr}', got invalid: {e.Message}");
        }
        catch (Exception e)
        {
            return VectorResult.Fail("consent", name,
                $"unexpected exception type: {e.GetType().Name}: {e.Message}");
        }
    }

    private static VectorResult RunRoundtrip(string sharedKey, JsonObject v)
    {
        string name = "roundtrip/" + v["name"]!.GetValue<string>();
        var issueSpec = v["issue"]!.AsObject();
        long ttl = issueSpec["ttl_seconds"]?.GetValue<long>() ?? 3600;
        string subject = issueSpec["subject"]!.GetValue<string>();
        var scopes = new List<string>();
        foreach (var n in issueSpec["scopes"]!.AsArray()) scopes.Add(n!.GetValue<string>());
        string? execId = StringOrDefault(issueSpec, "execution_id", null);
        var metadata = issueSpec["metadata"] as JsonObject;

        long now = v["now"] is JsonValue nv && nv.TryGetValue<long>(out var nt) ? nt : 1_700_000_000L;
        string issuer = StringOrDefault(v, "issuer", "iaiso") ?? "iaiso";
        var alg = ParseAlg(v);

        var iss = Issuer.CreateBuilder()
            .WithHsKey(Encoding.UTF8.GetBytes(sharedKey))
            .WithAlgorithm(alg)
            .WithIssuer(issuer)
            .WithClock(() => now)
            .Build();
        Scope scope;
        try { scope = iss.Issue(subject, scopes, execId, ttl, metadata); }
        catch (Exception e)
        {
            return VectorResult.Fail("consent", name, "issue failed: " + e.Message);
        }

        bool expectSuccess = v["expected_after_verify_succeeds"]?.GetValue<bool>() ?? false;
        string? verifyExec = StringOrDefault(v, "verify_with_execution_id", null);

        var ver = Verifier.CreateBuilder()
            .WithHsKey(Encoding.UTF8.GetBytes(sharedKey))
            .WithAlgorithm(alg)
            .WithIssuer(issuer)
            .WithClock(() => now + 1)
            .Build();
        try
        {
            var verified = ver.Verify(scope.Token, verifyExec);
            if (!expectSuccess)
                return VectorResult.Fail("consent", name, "expected verify to fail, succeeded");
            if (verified.Subject != subject)
                return VectorResult.Fail("consent", name, "subject mismatch");
            if (!ListEquals(verified.Scopes, scopes))
                return VectorResult.Fail("consent", name, "scopes mismatch");
            return VectorResult.Pass("consent", name);
        }
        catch (Exception e)
        {
            return expectSuccess
                ? VectorResult.Fail("consent", name, "expected verify to succeed, failed: " + e.Message)
                : VectorResult.Pass("consent", name);
        }
    }

    private static bool ListEquals(IReadOnlyList<string> a, IReadOnlyList<string> b)
    {
        if (a.Count != b.Count) return false;
        for (int i = 0; i < a.Count; i++) if (a[i] != b[i]) return false;
        return true;
    }
}

File: IAIso-v5.0/core/iaiso-php/src/Conformance/ConsentRunner.php
<?php

declare(strict_types=1);

namespace IAIso\Conformance;

use IAIso\Consent\Algorithm;
use IAIso\Consent\ExpiredTokenException;
use IAIso\Consent\InvalidTokenException;
use IAIso\Consent\Issuer;
use IAIso\Consent\RevokedTokenException;
use IAIso\Consent\Scopes;
use IAIso\Consent\Verifier;

/** @internal */
final class ConsentRunner
{
    private function __construct() {}

    /** @return VectorResult[] */
    public static function run(string $specRoot): array
    {
        $file = $specRoot . '/consent/vectors.json';
        $doc = json_decode((string) file_get_contents($file), true, 512, JSON_THROW_ON_ERROR);
        $sharedKey = $doc['hs256_key_shared'];

        $out = [];
        foreach ($doc['scope_match'] ?? [] as $v) {
            $out[] = self::runScopeMatch($v);
        }
        foreach ($doc['scope_match_errors'] ?? [] as $v) {
            $out[] = self::runScopeMatchError($v);
        }
        foreach ($doc['valid_tokens'] ?? [] as $v) {
            $out[] = self::runValidToken($sharedKey, $v);
        }
        foreach ($doc['invalid_tokens'] ?? [] as $v) {
            $out[] = self::runInvalidToken($sharedKey, $v);
        }
        foreach ($doc['issue_and_verify_roundtrip'] ?? [] as $v) {
            $out[] = self::runRoundtrip($sharedKey, $v);
        }
        return $out;
    }

    private static function runScopeMatch(array $v): VectorResult
    {
        $name = 'scope_match/' . $v['name'];
        try {
            $got = Scopes::granted($v['granted'], $v['requested']);
            $want = (bool) $v['expected'];
            if ($got !== $want) {
                return VectorResult::fail('consent', $name,
                    'got ' . ($got ? 'true' : 'false')
                    . ', want ' . ($want ? 'true' : 'false'));
            }
            return VectorResult::pass('consent', $name);
        } catch (\Throwable $e) {
            return VectorResult::fail('consent', $name,
                'unexpected exception: ' . $e->getMessage());
        }
    }

    private static function runScopeMatchError(array $v): VectorResult
    {
        $name = 'scope_match_errors/' . $v['name'];
        $expectErr = $v['expect_error'];
        try {
            Scopes::granted($v['granted'], $v['requested']);
            return VectorResult::fail('consent', $name,
                "expected error containing '$expectErr', got Ok");
        } catch (\InvalidArgumentException $e) {
            if (!str_contains(strtolower($e->getMessage()), strtolower($expectErr))) {
                return VectorResult::fail('consent', $name,
                    "expected '$expectErr', got: " . $e->getMessage());
            }
            return VectorResult::pass('consent', $name);
        }
    }

    private static function parseAlg(array $v): Algorithm
    {
        if (isset($v['algorithm'])) {
            return Algorithm::from($v['algorithm']);
        }
        return Algorithm::HS256;
    }

    private static function runValidToken(string $sharedKey, array $v): VectorResult
    {
        $name = 'valid_tokens/' . $v['name'];
        $now = (int) $v['now'];
        $issuer = $v['issuer'] ?? 'iaiso';
        $alg = self::parseAlg($v);

        $verifier = Verifier::builder()
            ->hsKey($sharedKey)->algorithm($alg)->issuer($issuer)
            ->clock(fn() => $now)->build();
        try {
            $s = $verifier->verify($v['token']);
            $exp = $v['expected'];
            if ($s->subject !== $exp['sub']) {
                return VectorResult::fail('consent', $name,
                    "sub: got {$s->subject}, want {$exp['sub']}");
            }
            if ($s->jti !== $exp['jti']) {
                return VectorResult::fail('consent', $name,
                    "jti: got {$s->jti}, want {$exp['jti']}");
            }
            $wantScopes = $exp['scopes'];
            if ($s->scopes !== $wantScopes) {
                return VectorResult::fail('consent', $name, 'scopes mismatch');
            }
            $wantExec = $exp['execution_id'] ?? null;
            if ($s->executionId !== $wantExec) {
                $g = $s->executionId ?? 'null';
                $w = $wantExec ?? 'null';
                return VectorResult::fail('consent', $name,
                    "execution_id: got $g, want $w");
            }
            return VectorResult::pass('consent', $name);
        } catch (\Throwable $e) {
            return VectorResult::fail('consent', $name, 'verify failed: ' . $e->getMessage());
        }
    }

    private static function runInvalidToken(string $sharedKey, array $v): VectorResult
    {
        $name = 'invalid_tokens/' . $v['name'];
        $now = (int) $v['now'];
        $issuer = $v['issuer'] ?? 'iaiso';
        $alg = self::parseAlg($v);
        $execId = $v['execution_id'] ?? null;
        $expectErr = $v['expect_error'];

        $verifier = Verifier::builder()
            ->hsKey($sharedKey)->algorithm($alg)->issuer($issuer)
            ->clock(fn() => $now)->build();
        try {
            $verifier->verify($v['token'], $execId);
            return VectorResult::fail('consent', $name,
                "expected error '$expectErr', got Ok");
        } catch (ExpiredTokenException) {
            if ($expectErr !== 'expired') {
                return VectorResult::fail('consent', $name, "expected '$expectErr', got expired");
            }
            return VectorResult::pass('consent', $name);
        } catch (RevokedTokenException) {
            if ($expectErr !== 'revoked') {
                return VectorResult::fail('consent', $name, "expected '$expectErr', got revoked");
            }
            return VectorResult::pass('consent', $name);
        } catch (InvalidTokenException $e) {
            if ($expectErr !== 'invalid') {
                return VectorResult::fail('consent', $name,
                    "expected '$expectErr', got invalid: " . $e->getMessage());
            }
            return VectorResult::pass('consent', $name);
        } catch (\Throwable $e) {
            return VectorResult::fail('consent', $name,
                'unexpected exception type: ' . get_class($e) . ': ' . $e->getMessage());
        }
    }

    private static function runRoundtrip(string $sharedKey, array $v): VectorResult
    {
        $name = 'roundtrip/' . $v['name'];
        $issueSpec = $v['issue'];
        $ttl = (int) ($issueSpec['ttl_seconds'] ?? 3600);
        $subject = $issueSpec['subject'];
        $scopes = $issueSpec['scopes'];
        $execId = $issueSpec['execution_id'] ?? null;
        $metadata = $issueSpec['metadata'] ?? null;
        $now = (int) ($v['now'] ?? 1700000000);
        $issuer = $v['issuer'] ?? 'iaiso';
        $alg = self::parseAlg($v);

        $is = Issuer::builder()->hsKey($sharedKey)->algorithm($alg)
            ->issuer($issuer)->clock(fn() => $now)->build();
        try {
            $scope = $is->issue($subject, $scopes, $execId, $ttl, $metadata);
        } catch (\Throwable $e) {
            return VectorResult::fail('consent', $name, 'issue failed: ' . $e->getMessage());
        }

        $expectSuccess = (bool) ($v['expected_after_verify_succeeds'] ?? false);
        $verifyExec = $v['verify_with_execution_id'] ?? null;

        $ver = Verifier::builder()->hsKey($sharedKey)->algorithm($alg)
            ->issuer($issuer)->clock(fn() => $now + 1)->build();
        try {
            $verified = $ver->verify($scope->token, $verifyExec);
            if (!$expectSuccess) {
                return VectorResult::fail('consent', $name, 'expected verify to fail, succeeded');
            }
            if ($verified->subject !== $subject) {
                return VectorResult::fail('consent', $name, 'subject mismatch');
            }
            if ($verified->scopes !== $scopes) {
                return VectorResult::fail('consent', $name, 'scopes mismatch');
            }
            return VectorResult::pass('consent', $name);
        } catch (\Throwable $e) {
            if ($expectSuccess) {
                return VectorResult::fail('consent', $name,
                    'expected verify to succeed, failed: ' . $e->getMessage());
            }
            return VectorResult::pass('consent', $name);
        }
    }
}
